README.md
# Anonlang #

The top priority for this language is developer productivity, and more specifically, quickly getting ideas to implementations and running. Another top focus is typing speed, comfort, and flow, more about that later.

Developers should prefer using the high-level abstractions, but the low-levels are fully accessible if needed.

Some describe Anonlang as having the power of Lisp with readability elegance of Ruby.

<!-- Anonlang: An abstraction of language, that you can program in. -->



## Sample Code for an App ##

    config [
      name 'My Minimal App'
    ]
    alias root layout [
      text 'Hello, World!'
    ]
    show-window root



## Top Reasons To Use Anonlang ##

### A joy to program in ###
Typing comfort is a huge part of Anonlang. There is a minimal need of using the shift-key in the core language. To define blocks/scopes, the square brackets are used instead of parenthesis or curly braces. This makes sense considering how much they are used in every program. Similar decision have been made, where the more often used features are easier to type than lesser used features.

### Quick to learn and get started ###
There are just a few general rules to the language. Also, it's just a small download to run Anonlang programs directly on your computer.


- Three data types:
  - `boolean` which can be `true` or `false`
  - `number` integer or floating, this keeps track of sig figs used
  - `text` is always within single quotes, ex: `'my text'`
    - A variable can be included in text by surrounding it with semi-colons, ex: `hi ;name;!`]
    - Concat text using `++`

- Single-line comments start with `;;`
- Multi-line comments are between `[;` and `;]`


#### Keywords ####

    alias, loop

    Examples:

    alias get-current-time date.now.time ;; Create lazy pointer to something without running it
    alias nodes node-group [
      node [ 'en', 'hello' ]
      node [ 'es', 'hola' ]
      node [ 'jp', 'konnichiwa' ]
    ]

    loop [ ... ] ;; Loop forever
    loop 5 [ ... ] ;; Loop 5 times
    loop count in 1..5 [ output count ] ;; Loop with reference to iteration
    loop node in nodes [ visit node ] ;; Loop over each item
    loop ch in 'hello' [ output ch ] ;; Loop over each character
    loop wd in 'hello world' by r'\w' [ out wd ] ;; Loop over word tokens
    loop is-wait [ sleep 500 ] ;; Loop until is-wait is false, aka while is-wait is true

    loop-parallel [] ;; When order of calls doesn't matter.



## App Lifecycle ##

Apps have a simple life-cycle:

    start
      |
      V
    stop

Apps with GUI have:

    start
      |
      V
    show <--+
      |     |
      V     |
    hide ---+
      |
      V
    stop

Though, simple apps (or scripts) don't need to worry about that:

    ;; A simple app
    input a
    output 'Hello, ;a;!'



## Paradigm ##
Anonlang is not based on any language, but it does share many commonalities with others.

### Everything is a Everything ###

- A variable is a function with zero inputs and one output.
- A function is a class with code on initialization and optional output. Or, another way, a class is a function that returns itself.
- Classes connecting to one another is the basis for a graph data structure, which in turn is the base for trees and linked lists. Arrays could be simulated with a graph or linked list, not paying attention to performance.

These are all first-class citizens in Anonlang.

### What If? ###

- What if there was a programming language where you don't have to worry about performance? Quite different without many different data structures and algorithms?
- What if there was a programming language where the language was designed after the modern standard keyboard layout?
  - Use square bracket more than curly brace and parenthesis?
- What if there was a programming language that had a minimum number of tokens to type? What would it look like? Likely, somewhat declarative.



## Background ##
This is a 10-year project to learn and re-think and re-implement programming. All assumptions will be questioned without bias.

Some ideas about possible generalized axioms:

- Everything has an input and output.





<!-- Design Philosophy: Re-using common syntax is NOT a requirement. -->


changelog.txt


2018-04-25: Don't need to use `alias` or `define` to create a new type because in small programs, it is just a lot of duplicate code.



compiler/in-java/src/com/danialgoodwin/Main.java
package com.danialgoodwin;

import com.anonsage.anonlang.AnonlangAst;
import com.anonsage.anonlang.AnonlangAstToHtml;
import com.anonsage.anonlang.AnonlangAstToJson;
import com.anonsage.artlang.ArtlangLexer;
import com.anonsage.artlang.ArtlangParser;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Scanner;

public class Main {

    private static final boolean IS_READ_FROM_FILE = true;

    public static void main(String[] args) {
        log("main()");
        ArtlangLexer lexer = new ArtlangLexer();
        if (IS_READ_FROM_FILE) {
            try (BufferedReader br = new BufferedReader(new FileReader("sample.txt"))) {
                String line;
                while ((line = br.readLine()) != null) {
                    if (line.isEmpty()) { break; }
                    lexer.add(line);
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            Scanner scanner = new Scanner(System.in);
            while (true) {
                String line = scanner.nextLine();
                if (line.isEmpty()) { break; }
                lexer.add(line);
            }
        }
        ArrayList<AnonlangAst.Token> tokens = lexer.getTokens();
        log("tokens=" + tokens);

        ArtlangParser parser = new ArtlangParser();
        parser.add(tokens);
        AnonlangAst ast = parser.getAst();
        log("ast=" + ast);

        String html = AnonlangAstToHtml.convert(ast);
        log("html=\n" + html);

        String json = AnonlangAstToJson.convert(ast);
        log("json=\n" + json);
    }

    public static void log(String message) {
        System.out.println(message);
    }

}


compiler/in-java/src/com/anonsage/anonlang/AnonlangAst.java
package com.anonsage.anonlang;

import java.util.List;

/**
 * 2016-05-28: Created by Danial Goodwin
 */
public class AnonlangAst {

    public static class ProgNode extends AnonlangAst {
        List<GroupNode> groups;
        public ProgNode(List<GroupNode> groups) {
            this.groups = groups;
        }
        @Override
        public String toString() {
            return "ProgNode{" +
                    "groups=" + groups +
                    '}';
        }
    }

    public static class GroupNode extends AnonlangAst {
        List<AttributeNode> attributes;
        public GroupNode(List<AttributeNode> attributes) {
            this.attributes = attributes;
        }
        @Override
        public String toString() {
            return "GroupNode{" +
                    "attributes=" + attributes +
                    '}';
        }
    }

    public static class AttributeNode extends AnonlangAst {
        final TextNode textNode;
        final BlockNode blockNode;
        public AttributeNode(TextNode textNode, BlockNode blockNode) {
            this.textNode = textNode;
            this.blockNode = blockNode;
        }
        @Override
        public String toString() {
            return "AttributeNode{" +
                    "textNode=" + textNode +
                    ", blockNode=" + blockNode +
                    '}';
        }
    }

    public static class BlockNode extends AnonlangAst {
        final AnonlangAst ast;
        public BlockNode(AnonlangAst ast) {
            this.ast = ast;
        }
        @Override
        public String toString() {
            return "BlockNode{" +
                    "ast=" + ast +
                    '}';
        }
    }

    public static class NumberNode extends AnonlangAst {
        final Token token;
        public NumberNode(Token token) {
            this.token = token;
        }
        @Override
        public String toString() {
            return "NumberNode{" +
                    "token=" + token +
                    '}';
        }
    }

    public static class TextNode extends AnonlangAst {
        final Token token;
        public TextNode(Token token) {
            this.token = token;
        }
        @Override
        public String toString() {
            return "TextNode{" +
                    "token=" + token +
                    '}';
        }
    }

    public static class Token {
//        public static final int WHITESPACE = 0;
        public static final int NUMBER = 1;
//        public static final int PLUS = 2;
//        public static final int DASH = 3;
//        public static final int ASTERISK = 4;
//        public static final int SLASH_FORWARD = 5;
//        public static final int SLASH_BACK = 6;
        public static final int TEXT = 7;
        public static final int BRACKET_START = 8;
        public static final int BRACKET_STOP = 9;
//        public static final int EQUALS = 10;
        public static final int COLON = 11;
        public static final int SINGLE_QUOTE = 12;
        public static final int COMMA = 13;
        public static final int PERIOD = 14;
        public static final int END_OF_LINE = 15;
        public static final int END_OF_FILE = 16;
        public final int type;
        public final String value;
        public Token(int type, String value) {
            this.type = type;
            this.value = value;
        }
        @Override
        public String toString() {
            return "Token{" +
                    "type=" + type +
                    ", value='" + value + '\'' +
                    '}';
        }
    }

}


compiler/in-java/src/com/anonsage/artlang/ArtlangLexer.java
/**
 * 2016-05-28: Created by Danial Goodwin
 */
package com.anonsage.artlang;

import com.anonsage.anonlang.AnonlangAst;

import java.util.ArrayList;

/**
 *
 * grammar: [
 *   prog: group (group)*
 *   group: attribute (attribute)*
 *   attribute: TEXT COLON block
 *   block: NUMBER | TEXT | (BRACKET_START group BRACKET_STOP)
 * ]
 */
public class ArtlangLexer {

    private String currentLine;
    private int currentPosition;
    private char currentChar;

    private ArrayList<AnonlangAst.Token> tokens = new ArrayList<>();

    public void add(String currentLine) {
        this.currentLine = currentLine;
        currentPosition = 0;
        currentChar = currentLine.charAt(currentPosition);
        while (currentPosition < currentLine.length()) {
            AnonlangAst.Token token = getNextToken();
            tokens.add(token);
        }
    }

    public ArrayList<AnonlangAst.Token> getTokens() {
        return tokens;
    }

    private boolean moveToNextChar() {
        currentPosition++;
        if (currentPosition >= currentLine.length()) {
            return false;
        } else {
            currentChar = currentLine.charAt(currentPosition);
            return true;
        }
    }

    private AnonlangAst.Token getNextToken() {
        //noinspection StatementWithEmptyBody
        while (Character.isWhitespace(currentChar) && moveToNextChar()) {}
        if (currentPosition >= currentLine.length()) { return new AnonlangAst.Token(AnonlangAst.Token.END_OF_LINE, ""); }
        if (currentChar == ':') { moveToNextChar(); return new AnonlangAst.Token(AnonlangAst.Token.COLON, ":"); }
        if (currentChar == '[') { moveToNextChar(); return new AnonlangAst.Token(AnonlangAst.Token.BRACKET_START, "["); }
        if (currentChar == ']') { moveToNextChar(); return new AnonlangAst.Token(AnonlangAst.Token.BRACKET_STOP, "]"); }
        if (currentChar == '\'') { moveToNextChar(); return new AnonlangAst.Token(AnonlangAst.Token.SINGLE_QUOTE, "'"); }
        if (Character.isDigit(currentChar)) { return new AnonlangAst.Token(AnonlangAst.Token.NUMBER, number()); }
        if (Character.isLetterOrDigit(currentChar)) { return new AnonlangAst.Token(AnonlangAst.Token.TEXT, text()); }
        throw new RuntimeException("Unexpected character: " + currentChar);
    }

    private String number() {
        StringBuilder sb = new StringBuilder();
        while (Character.isDigit(currentChar)) {
            sb.append(currentChar);
            if (!moveToNextChar()) { break; }
        }
        return sb.toString();
    }

    private String text() {
        StringBuilder sb = new StringBuilder();
        while (Character.isLetterOrDigit(currentChar)) {
            sb.append(currentChar);
            if (!moveToNextChar()) { break; }
        }
        return sb.toString();
    }

}


compiler/in-java/src/com/anonsage/artlang/ArtlangParser.java
package com.anonsage.artlang;

import com.anonsage.anonlang.AnonlangAst;

import java.util.ArrayList;

/**
 * 2016-05-28: Created by Danial Goodwin
 */
public class ArtlangParser {

    private ArrayList<AnonlangAst.Token> tokens;
    private AnonlangAst ast;
    private int currentPosition;
    private AnonlangAst.Token currentToken;

    public void add(ArrayList<AnonlangAst.Token> tokens) {
        if (tokens == null || tokens.isEmpty()) { throw new RuntimeException("Input must be non-empty");}
        this.tokens = tokens;
        currentPosition = 0;
        currentToken = tokens.get(currentPosition);
        ast = prog();
    }

    public AnonlangAst getAst() {
        return ast;
    }

    private void moveToNextTokenIfMatch(int tokenType) {
        if (currentToken.type == tokenType) {
            currentPosition++;
            if (currentPosition >= tokens.size()) {
                currentToken = new AnonlangAst.Token(AnonlangAst.Token.END_OF_FILE, "");
            } else {
                currentToken = tokens.get(currentPosition);
            }
        } else {
            throw new RuntimeException("Unexpected type: " + tokenType);
        }
    }

    private AnonlangAst.ProgNode prog() {
        ArrayList<AnonlangAst.GroupNode> groups = new ArrayList<>();
        AnonlangAst.GroupNode group = group();
        groups.add(group);
        while (currentToken.type != AnonlangAst.Token.END_OF_FILE && (group = group()) != null) {
            groups.add(group);
        }
        return new AnonlangAst.ProgNode(groups);
    }

    private AnonlangAst.GroupNode group() {
        ArrayList<AnonlangAst.AttributeNode> attributes = new ArrayList<>();
        AnonlangAst.AttributeNode attribute = attribute();
        attributes.add(attribute);
        while (currentToken.type != AnonlangAst.Token.END_OF_FILE && currentToken.type != AnonlangAst.Token.BRACKET_STOP && (attribute = attribute()) != null) {
            attributes.add(attribute);
        }
        return new AnonlangAst.GroupNode(attributes);
    }

    private AnonlangAst.AttributeNode attribute() {
//        if (currentToken.type != AnonlangAst.Token.TEXT) { return null; }
        AnonlangAst.Token token = currentToken;
        moveToNextTokenIfMatch(AnonlangAst.Token.TEXT);
        moveToNextTokenIfMatch(AnonlangAst.Token.COLON);
        AnonlangAst.BlockNode block = block();
        return new AnonlangAst.AttributeNode(new AnonlangAst.TextNode(token), block);
    }

    private AnonlangAst.BlockNode block() {
        AnonlangAst.Token token = currentToken;
        switch (token.type) {
            case AnonlangAst.Token.NUMBER: moveToNextTokenIfMatch(AnonlangAst.Token.NUMBER); return new AnonlangAst.BlockNode(new AnonlangAst.NumberNode(token));
            case AnonlangAst.Token.TEXT: moveToNextTokenIfMatch(AnonlangAst.Token.TEXT); return new AnonlangAst.BlockNode(new AnonlangAst.TextNode(token));
            case AnonlangAst.Token.BRACKET_START:
                moveToNextTokenIfMatch(AnonlangAst.Token.BRACKET_START);
                AnonlangAst group = group();
                moveToNextTokenIfMatch(AnonlangAst.Token.BRACKET_STOP);
                return new AnonlangAst.BlockNode(group);
            default:
                throw new RuntimeException("Unexpected token for block: " + token);
        }
    }

}


compiler/in-java/src/com/anonsage/anonlang/AnonlangAstToHtml.java
package com.anonsage.anonlang;

/**
 * 2016-05-28: Created by Danial Goodwin
 */
public class AnonlangAstToHtml {

    public static String convert(AnonlangAst ast) {
        return visit(ast, 0);
    }

    private static String visit(AnonlangAst ast, int indentLevel) {
        if (ast instanceof AnonlangAst.ProgNode) {
            return visitProgNode((AnonlangAst.ProgNode) ast, indentLevel);
        } else if (ast instanceof AnonlangAst.GroupNode) {
            return visitGroupNode((AnonlangAst.GroupNode) ast, indentLevel);
        } else if (ast instanceof AnonlangAst.AttributeNode) {
            return visitAttributeNode((AnonlangAst.AttributeNode) ast, indentLevel);
        } else if (ast instanceof AnonlangAst.BlockNode) {
            return visitBlockNode((AnonlangAst.BlockNode) ast, indentLevel);
        } else if (ast instanceof AnonlangAst.TextNode) {
            return visitTextNode((AnonlangAst.TextNode) ast);
        } else if (ast instanceof AnonlangAst.NumberNode) {
            return visitNumberNode((AnonlangAst.NumberNode) ast);
        }
        throw new RuntimeException("Unexpected node: " + ast);
    }

    private static String visitProgNode(AnonlangAst.ProgNode ast, int indentLevel) {
        StringBuilder sb = new StringBuilder();
        sb.append("<html>\n");
        if (ast.groups == null) { return ""; }
        for (AnonlangAst.GroupNode group : ast.groups) {
            sb.append(visitGroupNode(group, indentLevel));
        }
        return sb.append("</html>").toString();
    }

    private static String visitGroupNode(AnonlangAst.GroupNode ast, int indentLevel) {
        StringBuilder sb = new StringBuilder();
        if (ast == null || ast.attributes == null) { return ""; }
        for (AnonlangAst.AttributeNode attribute : ast.attributes) {
            sb.append(visitAttributeNode(attribute, indentLevel + 1)).append("\n");
        }
        return sb.toString();
    }

    private static String visitAttributeNode(AnonlangAst.AttributeNode ast, int indentLevel) {
        StringBuilder sb = new StringBuilder();
        addSpaces(sb, indentLevel);
        sb.append("<").append(visitTextNode(ast.textNode)).append(">\n");
        sb.append(visitBlockNode(ast.blockNode, indentLevel));
        addSpaces(sb, indentLevel);
        sb.append("</").append(visitTextNode(ast.textNode)).append(">");
        return sb.toString();
    }

    private static String visitBlockNode(AnonlangAst.BlockNode ast, int indentLevel) {
        StringBuilder sb = new StringBuilder();
        if (ast.ast instanceof AnonlangAst.NumberNode) {
            addSpaces(sb, indentLevel + 1);
            sb.append(visitNumberNode((AnonlangAst.NumberNode) ast.ast)).append("\n");
        } else if (ast.ast instanceof AnonlangAst.TextNode) {
            addSpaces(sb, indentLevel + 1);
            sb.append(visitTextNode((AnonlangAst.TextNode) ast.ast)).append("\n");
        } else if (ast.ast instanceof AnonlangAst.GroupNode) {
            sb.append(visitGroupNode((AnonlangAst.GroupNode) ast.ast, indentLevel)).append("");
        } else {
            throw new RuntimeException("Unexpected ast: " + ast);
        }
        return sb.toString();
    }

    private static String visitTextNode(AnonlangAst.TextNode ast) {
        return ast.token.value;
    }

    private static String visitNumberNode(AnonlangAst.NumberNode ast) {
        return ast.token.value;
    }

    private static StringBuilder addSpaces(StringBuilder sb, int indentLevel) {
        int countSpaces = indentLevel * 2;
        for (int i = 0; i < countSpaces; i++) {
            sb.append(" ");
        }
        return sb;
    }

}


compiler/in-java/src/com/anonsage/anonlang/AnonlangAstToJson.java
package com.anonsage.anonlang;

/**
 * 2016-05-28: Created by Danial Goodwin
 */
/**
 *
 *
 head: [
 title: Hello
 ]
 body: [
 header: [
 text: Hi
 ]
 ]

 {
 "head" = {
 "title" = "Hello"
 },
 "body" = {
 "text" = "Hi"
 }
 }

 */
public class AnonlangAstToJson {

    public static String convert(AnonlangAst ast) {
        return visit(ast, 0);
    }

    private static String visit(AnonlangAst ast, int indentLevel) {
        if (ast instanceof AnonlangAst.ProgNode) {
            return visitProgNode((AnonlangAst.ProgNode) ast, indentLevel);
        } else if (ast instanceof AnonlangAst.GroupNode) {
            return visitGroupNode((AnonlangAst.GroupNode) ast, indentLevel);
        } else if (ast instanceof AnonlangAst.AttributeNode) {
            return visitAttributeNode((AnonlangAst.AttributeNode) ast, indentLevel);
        } else if (ast instanceof AnonlangAst.BlockNode) {
            return visitBlockNode((AnonlangAst.BlockNode) ast, indentLevel);
        } else if (ast instanceof AnonlangAst.TextNode) {
            return visitTextNode((AnonlangAst.TextNode) ast);
        } else if (ast instanceof AnonlangAst.NumberNode) {
            return visitNumberNode((AnonlangAst.NumberNode) ast);
        }
        throw new RuntimeException("Unexpected node: " + ast);
    }

    private static String visitProgNode(AnonlangAst.ProgNode ast, int indentLevel) {
        StringBuilder sb = new StringBuilder();
        sb.append("{\n");
        if (ast.groups == null) { return ""; }
        for (AnonlangAst.GroupNode group : ast.groups) {
            sb.append(visitGroupNode(group, indentLevel));
        }
        return sb.append("\n}").toString();
    }

    private static String visitGroupNode(AnonlangAst.GroupNode ast, int indentLevel) {
        StringBuilder sb = new StringBuilder();
        if (ast == null || ast.attributes == null) { return ""; }
        boolean isFirst = true;
        for (AnonlangAst.AttributeNode attribute : ast.attributes) {
            if (isFirst) {
                isFirst = false;
                sb.append("").append(visitAttributeNode(attribute, indentLevel + 1)).append("");
            } else {
                sb.append(",\n").append(visitAttributeNode(attribute, indentLevel + 1)).append("");
            }
        }
        return sb.toString();
    }

    private static String visitAttributeNode(AnonlangAst.AttributeNode ast, int indentLevel) {
        StringBuilder sb = new StringBuilder();
        addSpaces(sb, indentLevel);
        sb.append("\"").append(visitTextNode(ast.textNode)).append("\" = ");
        sb.append(visitBlockNode(ast.blockNode, indentLevel));
        return sb.toString();
    }

    private static String visitBlockNode(AnonlangAst.BlockNode ast, int indentLevel) {
        StringBuilder sb = new StringBuilder();
        if (ast.ast instanceof AnonlangAst.NumberNode) {
            sb.append(visitNumberNode((AnonlangAst.NumberNode) ast.ast)).append("");
        } else if (ast.ast instanceof AnonlangAst.TextNode) {
            sb.append("\"").append(visitTextNode((AnonlangAst.TextNode) ast.ast)).append("\"");
        } else if (ast.ast instanceof AnonlangAst.GroupNode) {
            sb.append("{\n").append(visitGroupNode((AnonlangAst.GroupNode) ast.ast, indentLevel)).append("\n");
            addSpaces(sb, indentLevel).append("}");
        } else {
            throw new RuntimeException("Unexpected ast: " + ast);
        }
        return sb.toString();
    }

    private static String visitTextNode(AnonlangAst.TextNode ast) {
        return ast.token.value;
    }

    private static String visitNumberNode(AnonlangAst.NumberNode ast) {
        return ast.token.value;
    }

    private static StringBuilder addSpaces(StringBuilder sb, int indentLevel) {
        int countSpaces = indentLevel * 2;
        for (int i = 0; i < countSpaces; i++) {
            sb.append(" ");
        }
        return sb;
    }

}
